### ISSUE TITLE
Allow test-build that does not modify C++ keywords

### ISSUE THREAD

**Issue by anokfireball** [openPMD/openPMD-api#175]
An option to disable invasive source modification (e.g. ```#define public private```) is needed to ensure wider compiler support. See #173 for reasoning.

**Comment by ax3l** [comment]
I just saw one can even trigger an error in GCC with invasive tests.
When including `<sstream>`, redefining private-public can lead to:
```
/usr/include/c++/6/sstream:300:7: error: ‘struct std::__cxx11::basic_stringbuf<_CharT, _Traits, _Alloc>::__xfer_bufptrs’ redeclared with different access
       struct __xfer_bufptrs
```
https://stackoverflow.com/questions/47839718/sstream-redeclared-with-public-access-compiler-error

**Comment by ax3l** [comment]
PR #322 should be affected by this (at least on my machine; update: [yep](https://travis-ci.org/openPMD/openPMD-api/builds/422515202)), so let's reopen this issue.

One solution I could think of, if we want to keep intrusive "private" tests, is to define a macro for those keywords in our classes where we actually want to change private to public:
```C++
#ifndef OPENPMD_private_testable
#   define OPENPMD_private_testable private
#endif

class openPMDsomething
{
public:
OPENPMD_private_testable:
};
```

Idea reference: https://stackoverflow.com/a/48624567/2719194

That keeps the impacted classes to an absolute minimum.

**Comment by ax3l** [comment]
Maybe we also want to make `INVASIVE_TESTS` by default to be *off* and only enable it in (non-Windows) CI. Since it's actively changing class signatures with potentially bad behavior, we should not ship this by default to users? :) Just my gut feeling to make deployment more predictable with unexpected archs and compilers.

**Comment by anokfireball** [comment]
Ugh, this sounds like an incoming mess.

Both proposals seem appropriate. The macro-ed version of ```private:``` does what we need wile keeping GCC at ease. I'd rather not include the rationale in the name of the macro, though. ```OPENPMD_private:``` should be enough.

Concerning the limitation of ```INVASIVE_TESTS=ON``` to CI: Sure. This issue was closed at a time where we did not have release versions, but only ```dev```, and this strategy seemed more appropriate. The code covered by this switch is likely not going to be platform- or compiler-dependent, anyway.
